evm/src/cpu/kernel/asm/core/access_lists.asm

```diff
@@ -0,0 +1,104 @@
+/// Access lists for addresses and storage keys.
+/// The access list is stored in an array, with the length of the array in front of it.
+/// For storage keys, the address and key are stored as two consecutive elements.
+/// The array is stored in the SEGMENT_ACCESSED_ADDRESSES segment for addresses and in the SEGMENT_ACCESSED_STORAGE_KEYS segment for storage keys.
+/// Searching and inserting is done by doing a linear search through the array.
+/// If the address/storage key isn't found in the array, it is inserted at the end.
+/// TODO: Look into using a more efficient data structure for the access lists.
+
+%macro insert_accessed_addresses
+    %stack (addr) -> (addr, %%after)
+    %jump(insert_accessed_addresses)
+%%after:
+    // stack: (addr in accessed_address)
+%endmacro
+
+%macro insert_accessed_addresses_no_return
+    %stack (addr) -> (addr, %%after)
+    %jump(insert_accessed_addresses)
+%%after:
+    // stack: (addr in accessed_address)
+    POP
+%endmacro
+
+/// Inserts the address into the access list if it is not already present.
+/// Return 1 if the address was inserted, 0 if it was already present.
+global insert_accessed_addresses:
+    // stack: addr, retdest
+    PUSH 0 %mload_current(@SEGMENT_ACCESSED_ADDRESSES)
```

Comment:
Since this a transaction-level data, I think it should use kernel memory rather than current context memory?

Reply:
Yes, though I'm still trying to figure out the details. I think writing everything to kernel memory doesn't quite work, because if a scope reverts, the access list should revert to its state before the scope was entered. So I think we need to also write the full access list in another place (e.g. another context) before any *CALL.

Reply:
Hm seems like you're right. I guess we could use context memory and descend the chain of parent contexts when checking if something is warm.

Reply:
(But what you have seems good for now, rollbacks in general aren't working yet)